src/dotnet-setversion/Program.cs

```diff
@@ -2,44 +2,107 @@
+            return Parser.Default.ParseArguments<Options>(args)
+                .MapResult(Run, errs => int.MinValue);
+        }
+
+        private static int Run(Options options)
+        {
+            if (options.Recursive && options.CsprojFile != null)
             {
-                Console.WriteLine("Missing version string");
-                return 1;
+                Console.WriteLine("The ---recurvice option and csprojFile argument are mutually exclusive.");
+                return -1;
             }
 
-            var versionString = args[0];
-            var currentDirectory = Directory.GetCurrentDirectory();
-            var csprojFile = Directory.EnumerateFileSystemEntries(currentDirectory, "*.csproj", SearchOption.TopDirectoryOnly).First();
-            var document = XDocument.Load(csprojFile);
+            return options.Recursive ? RunRecursive(options.Version) : Run(options.Version, options.CsprojFile);
+        }
 
-            // Search for an existing version node in the csproj file.
-            var projectNode = document.GetOrCreateElement("Project");
-            var versionNode = projectNode
-                .Elements("PropertyGroup")
-                .SelectMany(it => it.Elements("Version"))
-                .SingleOrDefault();
+        private static int Run(string version, string csprojFile)
+        {
+            if (csprojFile == null)
+            {
+                var csprojFiles = GetCsprojFiles(false);
+                if (!CheckCsprojFiles(csprojFiles, false, out var exitCode)) return exitCode;
+                csprojFile = csprojFiles[0];
+            }
 
-            // If no version node exists, create it.
-            if (versionNode == null)
+            if (!File.Exists(csprojFile))
             {
-                versionNode = projectNode
-                    .GetOrCreateElement("PropertyGroup")
-                    .GetOrCreateElement("Version");
+                Console.WriteLine($"Project file '{csprojFile}' does not exist.");
+                return -1009;
             }
 
-            versionNode.SetValue(versionString);
-            File.WriteAllText(csprojFile, document.ToString());
-            Console.WriteLine($"Setting version: {versionString}");
+            SetVersion(version, csprojFile);
+            PrintSuccessString(version, csprojFile);
             return 0;
         }
+
+        private static int RunRecursive(string version)
+        {
+            var csprojFiles = GetCsprojFiles(true);
+            if (!CheckCsprojFiles(csprojFiles, true, out var exitCode)) return exitCode;
+
+            foreach (var csprojFile in csprojFiles)
+            {
+                SetVersion(version, csprojFile);
+            }
+            PrintSuccessString(version, csprojFiles);
+            return 0;
+        }
+
+        private static string[] GetCsprojFiles(bool recursive)
+        {
+            return Directory.EnumerateFileSystemEntries(Directory.GetCurrentDirectory(), "*.csproj",
```

Comment:
Minor nitpick but I'd prefer to make this an expression-bodied method.

Reply:
Your codebase, your rules.